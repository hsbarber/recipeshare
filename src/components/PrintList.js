import React from 'react';
import PropTypes from 'prop-types';
import { DragDropContext, Droppable, Draggable } from 'react-beautiful-dnd';
import Remove from '../icons/Remove.js';

const getListStyle = isDraggingOver => ({
  background: isDraggingOver ? 'rgb(185, 185, 185)' : 'none',
});
const getItemStyle = (isDragging, draggableStyle) => ({
  // some basic styles to make the items look a bit nicer
  userSelect: 'none',

  // change background colour if dragging
  background: isDragging ? '#D0ACDC' : ' rgb(247, 247, 247)',

  // styles we need to apply on draggables
  ...draggableStyle,
});
function PrintList({ list, dragEnd, remove, name }) {
  return (
    <DragDropContext onDragEnd={dragEnd}>
      <Droppable droppableId="droppable">
        {(provided, snapshot) => (
          <ul
            ref={provided.innerRef}
            style={getListStyle(snapshot.isDraggingOver)}
            className="rForm--list"
          >
            {list.map((item, index) => (
              <Draggable key={item.id} draggableId={item.id} index={index}>
                {(provided, snapshot) => (
                  <li
                    ref={provided.innerRef}
                    {...provided.draggableProps}
                    {...provided.dragHandleProps}
                    style={getItemStyle(
                      snapshot.isDragging,
                      provided.draggableProps.style
                    )}
                  >
                    <span className="listNumber">{`${index + 1}.`}</span>
                    <span className="listContent">{item.content}</span>
                    <span
                      className="remove"
                      role="button"
                      tabIndex="0"
                      onKeyDown={() => {
                        remove(item.id, name);
                      }}
                      onClick={() => {
                        remove(item.id, name);
                      }}
                    >
                      <Remove color="#E24C3F" />
                    </span>
                  </li>
                )}
              </Draggable>
            ))}
            {provided.placeholder}
          </ul>
        )}
      </Droppable>
    </DragDropContext>
  );
}
export default PrintList;

PrintList.propTypes = {
  list: PropTypes.instanceOf(Array),
  dragEnd: PropTypes.func,
  remove: PropTypes.func,
  name: PropTypes.string,
};
