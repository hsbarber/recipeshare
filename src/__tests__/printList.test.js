import * as React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import PrintList from '../components/PrintList';
import '../enzyme';

const getDefaultProps = () => ({
  ingredients: [
    {
      content: 'eggs',
      id: 1552514776779,
    },
    {
      content: 'butter',
      id: 1552514776780,
    },
  ],
  dragEnd: jest.fn(),
  remove: jest.fn(),
});

describe('RecipeForm', () => {
  const { ingredients, dragEnd, remove } = getDefaultProps();
  it('shallow renders without crashing', () => {
    shallow(
      <PrintList
        list={ingredients}
        dragEnd={dragEnd}
        remove={remove}
        name="ingredients"
      />
    );
  });
  it('component matches snapshot', () => {
    const wrapper = shallow(
      <PrintList
        list={ingredients}
        dragEnd={dragEnd}
        remove={remove}
        name="ingredients"
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('renders a list based on ingredients props', () => {
    const wrapper = shallow(
      <PrintList
        list={ingredients}
        dragEnd={dragEnd}
        remove={remove}
        name="ingredients"
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('renders correct number of ingredients based on ingredients props', () => {
    const wrapper = mount(
      <PrintList
        list={ingredients}
        dragEnd={dragEnd}
        remove={remove}
        name="ingredients"
      />
    );
    expect(wrapper.find('.rForm--list').children()).toHaveLength(2);
  });
  it('remove function called on click of delete button', () => {
    const name = 'ingredients';
    const id = 1552514776779;
    const wrapper = mount(
      <PrintList
        list={ingredients}
        dragEnd={dragEnd}
        remove={remove}
        name="ingredients"
      />
    );
    expect(wrapper.find('.rForm--list').children()).toHaveLength(2);
    const firstLi = wrapper.find('.rForm--list li').first();
    firstLi.find('.remove').simulate('click', { id, name });
    expect(remove).toHaveBeenCalled();
  });
});
