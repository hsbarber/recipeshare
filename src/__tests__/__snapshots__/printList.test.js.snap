// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeForm component matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PrintList
    dragEnd={[MockFunction]}
    list={
      Array [
        Object {
          "content": "eggs",
          "id": 1552514776779,
        },
        Object {
          "content": "butter",
          "id": 1552514776780,
        },
      ]
    }
    name="ingredients"
    remove={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Connect(Droppable)
        direction="vertical"
        droppableId="droppable"
        ignoreContainerClipping={false}
        isDropDisabled={false}
        type="DEFAULT"
      >
        [Function]
      </Connect(Droppable)>,
      "onDragEnd": [MockFunction],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": [Function],
        "direction": "vertical",
        "droppableId": "droppable",
        "ignoreContainerClipping": false,
        "isDropDisabled": false,
        "type": "DEFAULT",
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Connect(Droppable)
          direction="vertical"
          droppableId="droppable"
          ignoreContainerClipping={false}
          isDropDisabled={false}
          type="DEFAULT"
        >
          [Function]
        </Connect(Droppable)>,
        "onDragEnd": [MockFunction],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": [Function],
          "direction": "vertical",
          "droppableId": "droppable",
          "ignoreContainerClipping": false,
          "isDropDisabled": false,
          "type": "DEFAULT",
        },
        "ref": null,
        "rendered": [Function],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`RecipeForm renders a list based on ingredients props 1`] = `
<DragDropContext
  onDragEnd={[MockFunction]}
>
  <Connect(Droppable)
    direction="vertical"
    droppableId="droppable"
    ignoreContainerClipping={false}
    isDropDisabled={false}
    type="DEFAULT"
  >
    <Component />
  </Connect(Droppable)>
</DragDropContext>
`;
